### PR TITLE
Add Drupal data source - Closes #582

### LINKED ISSUE #582: Add Drupal data source
### Description

Create a job that retrieves data from Drupal.

These jobs, registered during start, take care of regular database updates with source data, using data from 3rd party services. These updates are performed during regular intervals, according to the usage policy written in the terms of service.

The job should be identified as `newsfeed.retrieve.drupal`.

Convert the data to the format described in additional information section. Store retrieved data in Redis with its update date.

Refer to the table in additional information to get more details about the source endpoint.

### Motivation

A reliable data source is required to keep up with the blog service.

### Acceptance Criteria

- Data is retrieved, converted and stored in the database.
- Update intervals do not exceed those written in the terms of service.
- Relevant data is stored in the database.
- Add unit tests to Drupal-related functions.

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?
This PR resolves #582

### How was it solved?
- [x] Data is retrieved from Drupal.
- [x] Transform data according to specifications. 
- [x] Relevant data is stored in the database.
- [x] Added docker configuration for newsfeed service
- [x] Add unit tests to Drupal-related functions (normalizer).
- [x] Added integration test cases for retrieve news by params

### How was it tested?
- [x] Local
- [x] Jenkins

### CODE REVIEW

File: services/newsfeed/config_drupal.js [hunk 1]
```diff
@@ -15,51 +15,30 @@
  */
 const getDrupalConfig = ({ enabled, name, url, filter }) => ({
 	name,
-	type: 'rest',
 	enabled,
 	url,
-	interval: 5 * 60, // seconds
+	table: 'newsfeed',
```
Comment (sameersubudhi) [inline services/newsfeed/config_drupal.js:20]: Kindly consider maintaining the `interval` within the config and referring it within the job.

File: services/newsfeed/config_drupal.js [hunk 2]
```diff
@@ -15,51 +15,30 @@
...
 			hash: '=,string',
 			title: '=,string',
-			content: 'description,string',
+			content_orig: 'description,string',
+			content_t: '=,string',
 			url: 'link,string',
 			image_url: '=,string',
-			timestamp: '=',
+			ctime: '=,number',
+			mtime: 'ctime,number',
```
Comment (sameersubudhi) [inline services/newsfeed/config_drupal.js:32]: `mtime` cannot always be the same as `ctime`. `ctime` is the post creation time and `mtime` is the last modified time. They can be equal initially though.

File: services/newsfeed/tests/unit/normalizer.test.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+const { normalizeData, normalizeFunctions } = require('../../shared/normalizers');
+const { drupalData } = require('../constants/newsfeed');
+const { newsfeedArticleSchema } = require('../schemas/newsfeedArticle.schema');
+const config = require('../../config');
+
+describe('Test normalizers', () => {
+	it('Test normalizeData', async () => {
+		const result = await normalizeData(config.sources.drupal_lisk_announcements, drupalData);
+		expect(result).toBeInstanceOf(Array);
+		expect(result.length).toBe(2);
```
Comment (sameersubudhi) [inline services/newsfeed/tests/unit/normalizer.test.js:25]:
```suggestion
		expect(result.length).toBe(drupalData.length);
```

File: services/newsfeed/tests/unit/normalizer.test.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		result.forEach(article => newsfeedArticleSchema.validate(article));
+	});
+
+	it('Test authorParser', async () => {
+		let result = await normalizeFunctions.authorParser(drupalData[0].author);
+		expect(result).toBe('Lisk');
+
+		result = await normalizeFunctions.authorParser('admin');
+		expect(result).toBe('Lisk');
```
Comment (sameersubudhi) [inline services/newsfeed/tests/unit/normalizer.test.js:34]:
```suggestion
		expect(['Lisk', drupalData[0].author]).toContain(result);

		result = await normalizeFunctions.authorParser('admin');
		expect(['Lisk', drupalData[0].author]).toContain(result);
```

File: tests/integration/api_v2/http/newsfeed.test.js [hunk 5]
```diff
@@ -38,7 +38,7 @@ describe('Newsfeed API', () => {
 				const response = await api.get(`${endpoint}`);
 				expect(response).toMap(goodRequestSchema);
 				expect(response.data).toBeInstanceOf(Array);
-				expect(response.data.length).toBeGreaterThanOrEqual(1);
+				expect(response.data.length).toBeGreaterThanOrEqual(10);
```
Comment (sameersubudhi) [inline tests/integration/api_v2/http/newsfeed.test.js:41]:
```suggestion
				expect(response.data.length).toBeGreaterThanOrEqual(1);
				expect(response.data.length).toBeLessThanOrEqual(10);
```

File: tests/integration/api_v2/http/newsfeed.test.js [hunk 6]
```diff
@@ -48,6 +48,30 @@ describe('Newsfeed API', () => {
...
+		});
+
+		it('retrieve news by multiple params', async () => {
+			const response = await api.get(`${endpoint}?source=drupal_lisk_general,drupal_lisk_announcements`);
+			expect(response).toMap(goodRequestSchema);
+			expect(response.data).toBeInstanceOf(Array);
+			expect(response.data.length).toBeGreaterThanOrEqual(1);
+			response.data.forEach(news => {
+				expect(news).toMap(newsfeedSchema);
+				expect(news.source).toMatch(/^\b(?:(?:drupal_lisk(?:_general|_announcements)|),?)+\b$/);
```
Comment (sameersubudhi) [inline tests/integration/api_v2/http/newsfeed.test.js:70]:
```suggestion
				expect(news.source).toMatch(/^\bdrupal_lisk(?:_general|_announcements)\b$/);
```

File: tests/integration/api_v2/rpc/newsfeed.test.js [hunk 7]
```diff
@@ -48,6 +48,34 @@ describe('Method get.newsfeed.articles', () => {
...
+		it('retrieve news by multiple sources', async () => {
+			const response = await getNewsfeed({ source: 'drupal_lisk_general,drupal_lisk_announcements' });
+			expect(response).toMap(jsonRpcEnvelopeSchema);
+			const { result } = response;
+			expect(result.data).toBeInstanceOf(Array);
+			expect(result.data.length).toBeGreaterThanOrEqual(1);
+			expect(result.data.length).toBeLessThanOrEqual(10);
+			result.data.forEach(news => {
+				expect(news).toMap(newsfeedSchema);
+				expect(news.source).toMatch(/^\b(?:(?:drupal_lisk(?:_general|_announcements)|),?)+\b$/);
```
Comment (sameersubudhi) [inline tests/integration/api_v2/rpc/newsfeed.test.js:74]:
```suggestion
				expect(news.source).toMatch(/^\bdrupal_lisk(?:_general|_announcements)\b$/);
```

File: tests/schemas/newsfeed.schema.js [hunk 8]
```diff
@@ -22,7 +22,8 @@ const newsfeedArticleSchema = {
 	source: Joi.string().pattern(/^\b[a-z]+(?:_[a-z]+){1,2}\b$/).required(),
 	source_id: Joi.string().required(),
 	ctime: Joi.number().integer().positive().required(),
-	mtime: Joi.number().integer().positive().required(),
+	mtime: Joi.number().integer().positive().allow(null)
+		.required(),
```
Comment (sameersubudhi) [inline tests/schemas/newsfeed.schema.js:26]: Why do we need to allow `null` values?

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Please my observations.

File: tests/schemas/newsfeed.schema.js (hunk 8 above)
Comment (nagdahimanshu) [reply tests/schemas/newsfeed.schema.js]: There is no need to allow `null`. It was a mistake and I reverted it back.

File: services/newsfeed/tests/unit/normalizer.test.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+		expect(result.length).toBe(drupalData.length);
+		result.forEach(article => newsfeedArticleSchema.validate(article));
+	});
+
+	it('Test authorParser', async () => {
+		let result = await normalizeFunctions.authorParser(drupalData[0].author);
+		expect(['Lisk', drupalData[0].author]).toContain(result);
+
+		result = await normalizeFunctions.authorParser('admin');
+		expect(['Lisk', drupalData[0].author]).toContain(result);
```
Comment (sameersubudhi) [inline services/newsfeed/tests/unit/normalizer.test.js:34]: Kindly revert this change as `'admin'` should always result in `'Lisk'`.
```suggestion
		expect(result).toEqual('Lisk');
```

File: tests/integration/api_v2/http/newsfeed.test.js [hunk 10]
```diff
@@ -48,6 +49,30 @@ describe('Newsfeed API', () => {
 			}
 		});
 
+		it('retrieve news by param', async () => {
+			const response = await api.get(`${endpoint}?source=drupal_lisk_general`);
+			expect(response).toMap(goodRequestSchema);
+			expect(response.data).toBeInstanceOf(Array);
+			expect(response.data.length).toBeGreaterThanOrEqual(1);
```
Comment (sameersubudhi) [inline tests/integration/api_v2/http/newsfeed.test.js:56]:
```suggestion
			expect(response.data.length).toBeGreaterThanOrEqual(1);
			expect(response.data.length).toBeLessThanOrEqual(10);
```

File: tests/integration/api_v2/http/newsfeed.test.js [hunk 11]
```diff
@@ -48,6 +49,30 @@ describe('Newsfeed API', () => {
...
+				expect(news.source).toEqual('drupal_lisk_general');
+			});
+			expect(response.meta).toMap(metaSchema);
+		});
+
+		it('retrieve news by multiple params', async () => {
+			const response = await api.get(`${endpoint}?source=drupal_lisk_general,drupal_lisk_announcements`);
+			expect(response).toMap(goodRequestSchema);
+			expect(response.data).toBeInstanceOf(Array);
+			expect(response.data.length).toBeGreaterThanOrEqual(1);
```
Comment (sameersubudhi) [inline tests/integration/api_v2/http/newsfeed.test.js:68]:
```suggestion
			expect(response.data.length).toBeGreaterThanOrEqual(1);
			expect(response.data.length).toBeLessThanOrEqual(10);
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

**Review by sameersubudhi — APPROVED** [review]
Thank you for the changes. LGTM 👍

**Review by MichalTuleja — APPROVED** [review]
